bokeh/util/structure.py

```diff
@@ -0,0 +1,315 @@
+        self._prop_df = self._make_prop_dict()
+        self._graph_plot = self.make_graph_plot()
+        self._data_table = self._make_data_table()
+        self._graph_plot.title.text = "Structure of model type {} with id {}".format(
+            self._model.__class__.__name__, self._model.id
+        )
+        self._structure_graph = self._combined()
+
+    @property
+    def model(self):
+        """
+        The bokeh model consisting of the structure graph and the datatable
+        for the attributes. Can be passed to show or file_html. Self contained,
+        so remains interactive in a notebook or html file; no server needed.
+        """
+        return self._structure_graph
+
+    @property
+    def graph(self):
+        """
+        The networkx digraph of submodels of the given model.
+        """
+        return self._graph
+
+    @property
+    def property_df(self):
+        """
+        A pandas dataframe of all of the properties of the model with their
+        values, types, and docstrings.  The base information for the datatable.
+        """
+        return self._prop_df
+
+    def _make_graph(self, M):
+        """Builds a networkx DiGraph() G so that:
+           G.nodes are the submodels of M, with node attributes
+               - "model" giving the class name of of the submodel
+               - "id" giving the id of the submodel
+            An edge runs from a to b if the submodel b belongs to an attribute of
+            a
+
+        Args:
+            A bokeh model M
+        """
+        def test_condition(s, y, H):
+            answer1 = False
+            answer2 = False
+            answer3 = False
+            try:
+                answer1 = (s in getattr(H, y))
+            except TypeError:
+                pass
+            try:
+                answer2 = (s == getattr(H, y))
+            except TypeError:
+                pass
+            try:
+                answer3 = (s in getattr(H, y).values())
+            except (AttributeError, ValueError):
+                pass
+            return (answer1 | answer2 | answer3)
+
+        K = nx.DiGraph()
+        T = {}
+        for m in M.references():
+            T[m.id] = set([y.id for y in m.references()])
+
+        K.add_nodes_from(
+            [(x, {"model": M.select_one({"id": x}).__class__.__name__}) for x in T]
+        )
+        E = [(y, x) for x, y in permutations(T, 2) if T[x] <= T[y]]
+        K.add_edges_from(E)
+        dead_edges = []
+        for id in K.nodes:
+            H = M.select_one({"id": id})
+            for x in K.neighbors(id):
+                s = H.select_one({"id": x})
+                keep_edge = False
+                for y in H.properties():
+                    if test_condition(s, y, H):
+                        keep_edge = True
+                if not keep_edge:
+                    dead_edges.append((id, x))
+        K.remove_edges_from(dead_edges)
+        return K
+
+    def obj_props_to_df2(self, obj):
```

Comment:
does this need to be public API?

Reply:
No, I've made the whole class and its methods private.  Actually this function can be useful generally -- it organizes all the properties and their docstrings in a dataframe -- but for now it can be private.